crt/fork.c

```diff
@@ -0,0 +1,427 @@
+#ifdef MYST_PSEUDO_FORK
+#define hidden __attribute__((__visibility__("hidden")))
+
+#define _GNU_SOURCE
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <linux/futex.h>
+#include <pthread.h>
+#include <sched.h>
+#include <signal.h>
+#include <stddef.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/mman.h>
+#include <sys/syscall.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include <pthread_impl.h>
+
+#include <myst/setjmp.h>
+#include <myst/syscallext.h>
+
+static int _round_up(uint64_t x, uint64_t m, uint64_t* r)
+{
+    uint64_t t;
+
+    if (!r)
+        return -EINVAL;
+
+    /* prevent divide by zero */
+    if (m == 0)
+        return -ERANGE;
+
+    if (__builtin_add_overflow(x, m - 1, &t))
+        return -ERANGE;
+
+    if (__builtin_mul_overflow(t / m, m, r))
+        return -ERANGE;
+
+    return 0;
+}
+
+/* ATTN: does not work for Linux yet: SYS_arch_prctl */
```

Comment:
Is the comment correct? (that it does not work on Linux yet)? If so, we can copy and modify this function from ``kernel/fsbase.c``.

```C++
void myst_set_fsbase(void* p)
{
    if (__options.have_syscall_instruction)
    {
        const long n = SYS_arch_prctl;
        myst_syscall2(n, ARCH_SET_FS, (long)p);
    }
    else
    {
        __asm__ volatile("wrfsbase %0" ::"r"(p));
    }
}
```

We will need a new system call to tell us that it is safe to use the **SYSCALL** instruction (e.g., ``SYS_have_syscall_instruction``).

Reply:
Also we should discuss with @vtikoo who is working on ``arch_prctl`` to add ``SET_FSBASE`` possible in the future.

Reply:
Can `SYS_set_thread_area` be used here? Thats the syscall the musl loader is calling for setting fsbase for the main thread.

Reply:
you added this comment :)
that said I think we are going to have that syscall, but only on a read-basis. Therefore a have_syscall type call may be problematic if we only support part.